// @fluxerjs/ws/dist/index.mjs
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined") return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/WebSocketManager.ts
import { EventEmitter as EventEmitter2 } from "events";

// src/WebSocketShard.ts
import { EventEmitter } from "events";
import { GatewayOpcodes } from "@fluxerjs/types";

// src/utils/getWebSocket.ts
import ws from "ws";
var cached = null;
function getDefaultWebSocketSync() {
  if (cached) return cached;
  if (typeof globalThis.WebSocket !== "undefined") {
    cached = globalThis.WebSocket;
    return cached;
  }
  if (typeof __require === "function") {
    try {
      cached = __require("ws");
      return cached;
    } catch {
    }
  }
  throw new Error(
    'No WebSocket implementation. Use Node 22+, or run with CommonJS. The "ws" package is bundled with @fluxerjs/ws.'
  );
}
async function getDefaultWebSocket() {
  if (cached) return cached;
  if (typeof globalThis.WebSocket !== "undefined") {
    cached = globalThis.WebSocket;
    return cached;
  }
  if (typeof __require === "function") {
    try {
      cached = __require("ws");
      return cached;
    } catch {
    }
  }
  return ws;
}

// src/WebSocketShard.ts
var RECONNECT_INITIAL_MS = 1e3;
var RECONNECT_MAX_MS = 45e3;
var WebSocketShard = class extends EventEmitter {
  ws = null;
  options;
  heartbeatInterval = null;
  heartbeatAt = 0;
  /** True until we send a heartbeat; then false until we get HeartbeatAck. Avoids closing before first heartbeat. */
  lastHeartbeatAck = true;
  sessionId = null;
  seq = null;
  destroying = false;
  url;
  WS;
  /** Current reconnect delay in ms; resets on successful connect. */
  reconnectDelayMs = RECONNECT_INITIAL_MS;
  reconnectTimeout = null;
  scheduleReconnect() {
    if (this.destroying) return;
    if (this.reconnectTimeout !== null) return;
    const delay = Math.min(RECONNECT_MAX_MS, this.reconnectDelayMs * (0.75 + Math.random() * 0.5));
    this.reconnectDelayMs = Math.min(RECONNECT_MAX_MS, this.reconnectDelayMs * 1.5);
    this.debug(`Reconnecting in ${Math.round(delay)}ms\u2026`);
    this.reconnectTimeout = setTimeout(() => {
      this.reconnectTimeout = null;
      this.connect();
    }, delay);
  }
  constructor(options) {
    super();
    this.options = options;
    this.WS = options.WebSocket ?? getDefaultWebSocketSync();
    const version = options.version ?? "1";
    const params = new URLSearchParams({ v: version, encoding: "json" });
    this.url = `${options.url}?${params}`;
  }
  get id() {
    return this.options.shardId;
  }
  get status() {
    if (!this.ws) return 0;
    switch (this.ws.readyState) {
      case 0:
        return 1;
      // Connecting
      case 1:
        return 2;
      // Ready
      case 2:
        return 3;
      // Closing
      case 3:
        return 0;
      default:
        return 0;
    }
  }
  connect() {
    if (this.ws?.readyState === 1) return;
    this.destroying = false;
    this.debug("Connecting");
    const ws2 = new this.WS(this.url);
    this.ws = ws2;
    const handleMessage = (data) => {
      try {
        const str = typeof data === "string" ? data : data.toString();
        this.handlePayload(JSON.parse(str));
      } catch (err) {
        this.emit("error", err instanceof Error ? err : new Error(String(err)));
      }
    };
    const handleClose = (code) => {
      this.ws = null;
      this.stopHeartbeat();
      this.emit("close", code);
      this.debug(`Closed: ${code}`);
      if (!this.destroying && shouldReconnectOnClose(code)) {
        this.scheduleReconnect();
      }
    };
    const handleError = (err) => {
      const error = err instanceof Error ? err : new Error("WebSocket error");
      this.emit("error", error);
      if (!this.destroying && this.ws) {
        this.ws = null;
        this.stopHeartbeat();
        this.debug("Connection error; will retry\u2026");
        this.scheduleReconnect();
      }
    };
    const handleOpen = () => {
      this.debug("Socket open");
      this.reconnectDelayMs = RECONNECT_INITIAL_MS;
    };
    if (typeof ws2.addEventListener === "function") {
      ws2.addEventListener("open", handleOpen);
      ws2.addEventListener(
        "message",
        (e) => handleMessage(e.data)
      );
      ws2.addEventListener("close", (e) => handleClose(e.code));
      ws2.addEventListener("error", () => handleError(new Error("WebSocket error")));
    } else if (typeof ws2.on === "function") {
      ws2.on("open", handleOpen);
      ws2.on("message", (d) => handleMessage(d));
      ws2.on("close", (code) => handleClose(code ?? 1006));
      ws2.on("error", (err) => handleError(err));
    }
  }
  debug(message) {
    this.emit("debug", `[Shard ${this.id}] ${message}`);
  }
  handlePayload(payload) {
    switch (payload.op) {
      case GatewayOpcodes.Hello:
        this.handleHello(payload.d);
        break;
      case GatewayOpcodes.HeartbeatAck:
        this.lastHeartbeatAck = true;
        break;
      case GatewayOpcodes.Dispatch:
        if (payload.t === "READY") {
          const d = payload.d;
          this.sessionId = d.session_id;
          this.reconnectDelayMs = RECONNECT_INITIAL_MS;
          this.emit("ready", payload.d);
        } else if (payload.t === "RESUMED") {
          this.reconnectDelayMs = RECONNECT_INITIAL_MS;
          this.emit("resumed");
        }
        if (payload.s !== void 0) this.seq = payload.s;
        this.emit("dispatch", payload);
        break;
      case GatewayOpcodes.InvalidSession:
        this.debug(`Invalid session (d=${payload.d}), reconnecting`);
        this.sessionId = null;
        this.seq = null;
        setTimeout(() => this.connect(), 1e3 + Math.random() * 4e3);
        break;
      case GatewayOpcodes.Reconnect:
        this.debug("Reconnect requested");
        this.ws?.close(1e3);
        setTimeout(() => this.connect(), 100);
        break;
      default:
        break;
    }
  }
  handleHello(data) {
    const jitter = Math.random() * data.heartbeat_interval;
    this.heartbeatAt = Date.now() + jitter;
    this.startHeartbeat(data.heartbeat_interval);
    if (this.sessionId && this.seq !== null) {
      this.send({
        op: GatewayOpcodes.Resume,
        d: {
          token: this.options.token,
          session_id: this.sessionId,
          seq: this.seq
        }
      });
    } else {
      const identify = {
        token: this.options.token,
        intents: this.options.intents,
        properties: {
          os: process.platform ?? "unknown",
          browser: "fluxerjs",
          device: "fluxerjs"
        }
      };
      if (this.options.presence) identify.presence = this.options.presence;
      this.send({ op: GatewayOpcodes.Identify, d: identify });
    }
  }
  startHeartbeat(interval) {
    this.stopHeartbeat();
    this.lastHeartbeatAck = true;
    this.heartbeatInterval = setInterval(() => {
      if (!this.lastHeartbeatAck && this.seq !== null) {
        this.debug("Heartbeat ack missed; reconnecting");
        this.ws?.close(1e3);
        if (!this.destroying) this.scheduleReconnect();
        return;
      }
      this.lastHeartbeatAck = false;
      this.send({ op: GatewayOpcodes.Heartbeat, d: this.seq ?? null });
    }, interval);
  }
  stopHeartbeat() {
    if (this.heartbeatInterval) {
      clearInterval(this.heartbeatInterval);
      this.heartbeatInterval = null;
    }
  }
  send(payload) {
    if (this.ws?.readyState !== 1) return;
    this.ws.send(JSON.stringify(payload));
  }
  destroy() {
    this.destroying = true;
    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
      this.reconnectTimeout = null;
    }
    this.stopHeartbeat();
    this.ws?.close(1e3);
    this.ws = null;
    this.sessionId = null;
    this.seq = null;
  }
};
function shouldReconnectOnClose(code) {
  switch (code) {
    case 1e3:
    // Normal Closure (server may close with this to ask for reconnect)
    case 1001:
    // Going Away
    case 1011:
    // Internal Error (server hit unexpected condition; often transient)
    case 1005:
    // No Status Received
    case 1006:
    // Abnormal Closure (network drop, server crash)
    case 1012:
    // Service Restart
    case 1013:
    // Try Again Later
    case 1014:
    // Bad Gateway
    case 1015:
      return true;
    case 4e3:
    // Unknown error
    case 4007:
    // Invalid seq
    case 4009:
    // Session timeout
    case 4010:
    // Invalid shard
    case 4011:
    // Sharding required
    case 4012:
      return true;
    default:
      return false;
  }
}

// src/WebSocketManager.ts
var RETRY_INITIAL_MS = 1e3;
var RETRY_MAX_MS = 45e3;
function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
var WebSocketManager = class extends EventEmitter2 {
  options;
  shards = /* @__PURE__ */ new Map();
  gatewayUrl = null;
  shardCount = 1;
  _aborted = false;
  constructor(options) {
    super();
    this.options = options;
  }
  async connect() {
    this._aborted = false;
    let WS = this.options.WebSocket;
    let delayMs = RETRY_INITIAL_MS;
    if (!WS) {
      while (!this._aborted) {
        try {
          WS = await getDefaultWebSocket();
          break;
        } catch (err) {
          const e = err instanceof Error ? err : new Error(String(err));
          this.emit("error", { shardId: -1, error: e });
          await sleep(delayMs);
          delayMs = Math.min(RETRY_MAX_MS, Math.floor(delayMs * 1.5));
        }
      }
      if (this._aborted) throw new Error("Connection aborted");
      if (!WS) throw new Error("Failed to load WebSocket");
    }
    let gateway = null;
    while (!this._aborted) {
      try {
        gateway = await this.options.rest.get("/gateway/bot");
        break;
      } catch (err) {
        const e = err instanceof Error ? err : new Error(String(err));
        this.emit("error", { shardId: -1, error: e });
        await sleep(delayMs);
        delayMs = Math.min(RETRY_MAX_MS, Math.floor(delayMs * 1.5));
      }
    }
    if (this._aborted) throw new Error("Connection aborted");
    if (!gateway) throw new Error("Failed to fetch gateway");
    this.gatewayUrl = gateway.url;
    this.shardCount = this.options.shardCount ?? gateway.shards;
    const ids = this.options.shardIds ?? [...Array(this.shardCount).keys()];
    const version = this.options.version ?? "1";
    for (const id of ids) {
      const shard = new WebSocketShard({
        url: this.gatewayUrl ?? gateway.url,
        token: this.options.token,
        intents: this.options.intents,
        presence: this.options.presence,
        shardId: id,
        numShards: this.shardCount,
        version,
        WebSocket: WS
      });
      shard.on("ready", (data) => this.emit("ready", { shardId: id, data }));
      shard.on("resumed", () => this.emit("resumed", id));
      shard.on("dispatch", (payload) => this.emit("dispatch", { shardId: id, payload }));
      shard.on("close", (code) => this.emit("close", { shardId: id, code }));
      shard.on("error", (err) => this.emit("error", { shardId: id, error: err }));
      shard.on("debug", (msg) => this.emit("debug", msg));
      this.shards.set(id, shard);
      try {
        shard.connect();
      } catch (err) {
        this.emit("error", {
          shardId: id,
          error: err instanceof Error ? err : new Error(String(err))
        });
      }
    }
  }
  send(shardId, payload) {
    this.shards.get(shardId)?.send(payload);
  }
  destroy() {
    this._aborted = true;
    for (const shard of this.shards.values()) shard.destroy();
    this.shards.clear();
    this.gatewayUrl = null;
  }
  getShardCount() {
    return this.shardCount;
  }
};

// src/utils/constants.ts
var GatewayCloseCodes = {
  Normal: 1e3,
  GoingAway: 1001,
  ProtocolError: 1002,
  UnsupportedData: 1003,
  NoStatusReceived: 1005,
  AbnormalClosure: 1006,
  UnknownError: 4e3,
  UnknownOpcode: 4001,
  DecodeError: 4002,
  NotAuthenticated: 4003,
  AuthenticationFailed: 4004,
  AlreadyAuthenticated: 4005,
  InvalidSeq: 4007,
  RateLimited: 4008,
  SessionTimeout: 4009,
  InvalidShard: 4010,
  ShardingRequired: 4011,
  InvalidAPIVersion: 4012
};
export {
  GatewayCloseCodes,
  WebSocketManager,
  WebSocketShard,
  getDefaultWebSocket,
  getDefaultWebSocketSync
};
