// @fluxerjs/rest/dist/index.mjs
// src/REST.ts
import { EventEmitter } from "events";

// src/RateLimitManager.ts
var RateLimitManager = class {
  buckets = /* @__PURE__ */ new Map();
  globalResetAt = 0;
  getBucket(route) {
    return this.buckets.get(route);
  }
  setBucket(route, limit, remaining, resetAt) {
    this.buckets.set(route, { limit, remaining, resetAt });
  }
  setGlobalReset(resetAt) {
    this.globalResetAt = resetAt;
  }
  getGlobalReset() {
    return this.globalResetAt;
  }
  /** Returns ms to wait before we can send again (0 if no wait). */
  getWaitTime(route) {
    const now = Date.now();
    const globalWait = this.globalResetAt > now ? this.globalResetAt - now : 0;
    const bucket = this.buckets.get(route);
    const bucketWait = bucket && bucket.remaining <= 0 && bucket.resetAt > now ? bucket.resetAt - now : 0;
    return Math.max(globalWait, bucketWait);
  }
  /** Parse rate limit headers and update state. */
  updateFromHeaders(route, headers) {
    const limit = headers.get("X-RateLimit-Limit");
    const remaining = headers.get("X-RateLimit-Remaining");
    const reset = headers.get("X-RateLimit-Reset");
    if (limit !== null && remaining !== null && reset !== null) {
      const resetMs = parseInt(reset, 10);
      if (!Number.isNaN(resetMs)) {
        const resetAt = resetMs * 1e3;
        this.setBucket(route, parseInt(limit, 10), parseInt(remaining, 10), resetAt);
      }
    }
    const retryAfter = headers.get("Retry-After");
    if (retryAfter !== null) {
      const sec = parseInt(retryAfter, 10);
      const resetAt = Date.now() + (Number.isNaN(sec) ? 0 : sec * 1e3);
      this.setBucket(route, 1, 0, resetAt);
    }
  }
};

// src/errors/FluxerAPIError.ts
var FluxerAPIError = class _FluxerAPIError extends Error {
  code;
  statusCode;
  errors;
  constructor(body, statusCode) {
    super(body.message);
    this.name = "FluxerAPIError";
    this.code = body.code;
    this.statusCode = statusCode;
    this.errors = body.errors;
    Object.setPrototypeOf(this, _FluxerAPIError.prototype);
  }
  /** True if the error is retryable (429 rate limit, 5xx server errors). */
  get isRetryable() {
    return this.statusCode === 429 || this.statusCode >= 500 && this.statusCode < 600;
  }
};

// src/errors/RateLimitError.ts
var RateLimitError = class _RateLimitError extends FluxerAPIError {
  retryAfter;
  global;
  constructor(body, statusCode) {
    super(body, statusCode);
    this.retryAfter = body.retry_after;
    this.global = body.global ?? false;
    this.name = "RateLimitError";
    Object.setPrototypeOf(this, _RateLimitError.prototype);
  }
};

// src/errors/HTTPError.ts
var STATUS_MESSAGES = {
  502: "Bad Gateway \u2014 Fluxer API may be temporarily unavailable.",
  503: "Service Unavailable \u2014 Fluxer API is down or overloaded. Try again later.",
  504: "Gateway Timeout \u2014 Fluxer API did not respond in time."
};
var HTTPError = class _HTTPError extends Error {
  statusCode;
  body;
  constructor(statusCode, body) {
    const hint = STATUS_MESSAGES[statusCode];
    const detail = body?.trim() || (hint ?? "No body");
    super(`HTTP ${statusCode}: ${detail}`);
    this.name = "HTTPError";
    this.statusCode = statusCode;
    this.body = body;
    Object.setPrototypeOf(this, _HTTPError.prototype);
  }
  /** True if the error is retryable (429 rate limit, 5xx server errors). */
  get isRetryable() {
    return this.statusCode === 429 || this.statusCode >= 500 && this.statusCode < 600;
  }
};

// src/utils/files.ts
function toBlob(data) {
  if (data instanceof Blob) return data;
  return new Blob([data]);
}
function toFormDataFile(data, filename) {
  const blob = toBlob(data);
  if (typeof File !== "undefined") {
    return new File([blob], filename, { type: blob.type || "application/octet-stream" });
  }
  return blob;
}
function buildFormData(payloadJson, files) {
  const form = new FormData();
  const payload = { ...payloadJson };
  if (files?.length && !payload.attachments) {
    payload.attachments = files.map((f, i) => ({
      id: i,
      filename: f.filename ?? f.name
    }));
  }
  form.append("payload_json", JSON.stringify(payload));
  if (files?.length) {
    for (let i = 0; i < files.length; i++) {
      const file = files[i];
      if (!file) continue;
      const filename = file.filename ?? file.name;
      const part = toFormDataFile(file.data, filename);
      form.append(`files[${i}]`, part, filename);
    }
  }
  return form;
}

// src/RequestManager.ts
var RequestManager = class {
  token = null;
  options;
  rateLimiter = new RateLimitManager();
  constructor(options) {
    this.options = {
      api: options.api ?? "https://api.fluxer.app",
      version: options.version ?? "1",
      authPrefix: options.authPrefix ?? "Bot",
      timeout: options.timeout ?? 15e3,
      retries: options.retries ?? 3,
      userAgent: options.userAgent ?? "fluxerjs"
    };
  }
  setToken(token) {
    this.token = token;
  }
  get baseUrl() {
    return `${this.options.api}/v${this.options.version}`;
  }
  /** Hash route for rate limit bucket (use path without ids for grouping). */
  getRouteHash(route) {
    return route.replace(/\d{17,19}/g, ":id");
  }
  async waitForRateLimit(routeHash) {
    const wait = this.rateLimiter.getWaitTime(routeHash);
    if (wait > 0) await new Promise((r) => setTimeout(r, wait));
  }
  buildHeaders(_route, options, body) {
    const headers = {
      "User-Agent": this.options.userAgent,
      ...options.headers
    };
    if (options.auth !== false && this.token) {
      headers["Authorization"] = `${this.options.authPrefix} ${this.token}`;
    }
    if (body !== void 0 && !(body instanceof FormData)) {
      headers["Content-Type"] = "application/json";
    }
    return headers;
  }
  async request(method, route, options = {}) {
    const routeHash = this.getRouteHash(route);
    const url = route.startsWith("http") ? route : `${this.baseUrl}${route}`;
    await this.waitForRateLimit(routeHash);
    let body;
    if (options.body !== void 0) {
      if (options.body instanceof FormData) {
        body = options.body;
      } else if (options.files?.length && typeof options.body === "object" && options.body !== null) {
        body = buildFormData(options.body, options.files);
      } else {
        body = JSON.stringify(options.body);
      }
    }
    const headers = this.buildHeaders(route, options, body);
    let lastError = null;
    for (let attempt = 0; attempt <= this.options.retries; attempt++) {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.options.timeout);
      try {
        const response = await fetch(url, {
          method,
          headers,
          body,
          signal: controller.signal
        });
        this.rateLimiter.updateFromHeaders(routeHash, response.headers);
        if (response.status === 429) {
          const data = await response.json().catch(() => ({}));
          const retryAfter = (data.retry_after ?? parseInt(response.headers.get("Retry-After") ?? "0", 10)) * 1e3;
          this.rateLimiter.setBucket(routeHash, 1, 0, Date.now() + retryAfter);
          if (data.global) this.rateLimiter.setGlobalReset(Date.now() + retryAfter);
          throw new RateLimitError(
            {
              ...data,
              code: "RATE_LIMITED",
              message: data.message ?? "Rate limited",
              retry_after: data.retry_after ?? 0
            },
            response.status
          );
        }
        const text = await response.text();
        if (!response.ok) {
          let parsed;
          try {
            parsed = JSON.parse(text);
          } catch {
            throw new HTTPError(response.status, text);
          }
          throw new FluxerAPIError(parsed, response.status);
        }
        if (response.status === 204 || text.length === 0) return void 0;
        return JSON.parse(text);
      } catch (err) {
        const wrapped = err instanceof Error ? err : new Error(String(err));
        lastError = attempt > 0 ? new Error(`Retry ${attempt} failed: ${wrapped.message}`, {
          cause: wrapped
        }) : wrapped;
        if (err instanceof RateLimitError && attempt < this.options.retries) {
          const retryMs = err.retryAfter * 1e3;
          if (Number.isFinite(retryMs)) {
            await new Promise((r) => setTimeout(r, retryMs));
            continue;
          }
        }
        if (err instanceof FluxerAPIError || err instanceof HTTPError) throw err;
        if (attempt < this.options.retries) {
          await new Promise((r) => setTimeout(r, 500 * (attempt + 1)));
          continue;
        }
        throw lastError;
      } finally {
        clearTimeout(timeoutId);
      }
    }
    throw lastError ?? new Error("Request failed");
  }
};

// src/REST.ts
import { Routes } from "@fluxerjs/types";
var REST = class extends EventEmitter {
  requestManager;
  _token = null;
  constructor(options = {}) {
    super();
    this.requestManager = new RequestManager({
      api: options.api ?? "https://api.fluxer.app",
      version: options.version ?? "1",
      authPrefix: options.authPrefix ?? "Bot",
      timeout: options.timeout ?? 15e3,
      retries: options.retries ?? 3,
      userAgent: options.userAgent ?? "fluxerjs"
    });
  }
  /** Set the bot token for authenticated requests. */
  setToken(token) {
    this._token = token;
    this.requestManager.setToken(token);
    return this;
  }
  /** Current bot token, or null. */
  get token() {
    return this._token;
  }
  /** Send a GET request. */
  async get(route, options) {
    return this.requestManager.request("GET", route, { auth: options?.auth });
  }
  /** Send a POST request. */
  async post(route, options) {
    return this.requestManager.request("POST", route, {
      body: options?.body,
      auth: options?.auth,
      files: options?.files
    });
  }
  /** Send a PATCH request. */
  async patch(route, options) {
    return this.requestManager.request("PATCH", route, {
      body: options?.body,
      auth: options?.auth
    });
  }
  /** Send a PUT request. */
  async put(route, options) {
    return this.requestManager.request("PUT", route, {
      body: options?.body,
      auth: options?.auth
    });
  }
  /** Send a DELETE request. */
  async delete(route, options) {
    return this.requestManager.request("DELETE", route, { auth: options?.auth });
  }
  /** Route helpers (from @fluxerjs/types) for building paths. */
  static get Routes() {
    return Routes;
  }
};

// src/index.ts
import { Routes as Routes2 } from "@fluxerjs/types";
export {
  FluxerAPIError,
  HTTPError,
  REST,
  RateLimitError,
  RateLimitManager,
  RequestManager,
  Routes2 as Routes,
  buildFormData
};
