// @fluxerjs/types/dist/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var index_exports = {};
__export(index_exports, {
  APIErrorCode: () => APIErrorCode,
  ChannelType: () => ChannelType,
  DefaultMessageNotifications: () => DefaultMessageNotifications,
  EmbedMediaFlags: () => EmbedMediaFlags,
  FLUXER_EPOCH: () => FLUXER_EPOCH,
  GatewayDispatchEvents: () => GatewayDispatchEvents,
  GatewayOpcodes: () => GatewayOpcodes,
  GuildExplicitContentFilter: () => GuildExplicitContentFilter,
  GuildMFALevel: () => GuildMFALevel,
  GuildVerificationLevel: () => GuildVerificationLevel,
  MessageAttachmentFlags: () => MessageAttachmentFlags,
  MessageFlags: () => MessageFlags,
  MessageType: () => MessageType,
  OverwriteType: () => OverwriteType,
  Routes: () => Routes,
  WebhookType: () => WebhookType
});
module.exports = __toCommonJS(index_exports);

// src/common/snowflake.ts
var FLUXER_EPOCH = 14200704e5;

// src/api/channel.ts
var ChannelType = /* @__PURE__ */ ((ChannelType2) => {
  ChannelType2[ChannelType2["GuildText"] = 0] = "GuildText";
  ChannelType2[ChannelType2["DM"] = 1] = "DM";
  ChannelType2[ChannelType2["GuildVoice"] = 2] = "GuildVoice";
  ChannelType2[ChannelType2["GroupDM"] = 3] = "GroupDM";
  ChannelType2[ChannelType2["GuildCategory"] = 4] = "GuildCategory";
  ChannelType2[ChannelType2["GuildLink"] = 5] = "GuildLink";
  ChannelType2[ChannelType2["GuildLinkExtended"] = 998] = "GuildLinkExtended";
  return ChannelType2;
})(ChannelType || {});
var OverwriteType = /* @__PURE__ */ ((OverwriteType2) => {
  OverwriteType2[OverwriteType2["Role"] = 0] = "Role";
  OverwriteType2[OverwriteType2["Member"] = 1] = "Member";
  return OverwriteType2;
})(OverwriteType || {});

// src/api/guild.ts
var GuildVerificationLevel = /* @__PURE__ */ ((GuildVerificationLevel2) => {
  GuildVerificationLevel2[GuildVerificationLevel2["None"] = 0] = "None";
  GuildVerificationLevel2[GuildVerificationLevel2["Low"] = 1] = "Low";
  GuildVerificationLevel2[GuildVerificationLevel2["Medium"] = 2] = "Medium";
  GuildVerificationLevel2[GuildVerificationLevel2["High"] = 3] = "High";
  GuildVerificationLevel2[GuildVerificationLevel2["VeryHigh"] = 4] = "VeryHigh";
  return GuildVerificationLevel2;
})(GuildVerificationLevel || {});
var GuildMFALevel = /* @__PURE__ */ ((GuildMFALevel2) => {
  GuildMFALevel2[GuildMFALevel2["None"] = 0] = "None";
  GuildMFALevel2[GuildMFALevel2["Elevated"] = 1] = "Elevated";
  return GuildMFALevel2;
})(GuildMFALevel || {});
var GuildExplicitContentFilter = /* @__PURE__ */ ((GuildExplicitContentFilter2) => {
  GuildExplicitContentFilter2[GuildExplicitContentFilter2["Disabled"] = 0] = "Disabled";
  GuildExplicitContentFilter2[GuildExplicitContentFilter2["MembersWithoutRoles"] = 1] = "MembersWithoutRoles";
  GuildExplicitContentFilter2[GuildExplicitContentFilter2["AllMembers"] = 2] = "AllMembers";
  return GuildExplicitContentFilter2;
})(GuildExplicitContentFilter || {});
var DefaultMessageNotifications = /* @__PURE__ */ ((DefaultMessageNotifications2) => {
  DefaultMessageNotifications2[DefaultMessageNotifications2["AllMessages"] = 0] = "AllMessages";
  DefaultMessageNotifications2[DefaultMessageNotifications2["OnlyMentions"] = 1] = "OnlyMentions";
  return DefaultMessageNotifications2;
})(DefaultMessageNotifications || {});

// src/api/embed.ts
var EmbedMediaFlags = {
  /** Embed media contains explicit content */
  CONTAINS_EXPLICIT_MEDIA: 16,
  /** Embed media is animated */
  IS_ANIMATED: 32
};

// src/api/message.ts
var MessageType = /* @__PURE__ */ ((MessageType2) => {
  MessageType2[MessageType2["Default"] = 0] = "Default";
  MessageType2[MessageType2["RecipientAdd"] = 1] = "RecipientAdd";
  MessageType2[MessageType2["RecipientRemove"] = 2] = "RecipientRemove";
  MessageType2[MessageType2["Call"] = 3] = "Call";
  MessageType2[MessageType2["ChannelNameChange"] = 4] = "ChannelNameChange";
  MessageType2[MessageType2["ChannelIconChange"] = 5] = "ChannelIconChange";
  MessageType2[MessageType2["ChannelPinnedMessage"] = 6] = "ChannelPinnedMessage";
  MessageType2[MessageType2["UserJoin"] = 7] = "UserJoin";
  MessageType2[MessageType2["Reply"] = 19] = "Reply";
  return MessageType2;
})(MessageType || {});
var MessageFlags = /* @__PURE__ */ ((MessageFlags2) => {
  MessageFlags2[MessageFlags2["SuppressEmbeds"] = 4] = "SuppressEmbeds";
  MessageFlags2[MessageFlags2["SuppressNotifications"] = 4096] = "SuppressNotifications";
  MessageFlags2[MessageFlags2["VoiceMessage"] = 8192] = "VoiceMessage";
  MessageFlags2[MessageFlags2["CompactAttachments"] = 131072] = "CompactAttachments";
  return MessageFlags2;
})(MessageFlags || {});
var MessageAttachmentFlags = {
  /** Renders attachment as spoiler (blurred until clicked) */
  IS_SPOILER: 8,
  /** Attachment contains explicit content */
  CONTAINS_EXPLICIT_MEDIA: 16,
  /** Animated image (e.g. GIF, animated WebP). Renders with EmbedGif / looping behavior. */
  IS_ANIMATED: 32
};

// src/api/webhook.ts
var WebhookType = /* @__PURE__ */ ((WebhookType2) => {
  WebhookType2[WebhookType2["Incoming"] = 1] = "Incoming";
  WebhookType2[WebhookType2["ChannelFollower"] = 2] = "ChannelFollower";
  return WebhookType2;
})(WebhookType || {});

// src/api/errors.ts
var APIErrorCode = /* @__PURE__ */ ((APIErrorCode2) => {
  APIErrorCode2["Unauthorized"] = "UNAUTHORIZED";
  APIErrorCode2["Forbidden"] = "FORBIDDEN";
  APIErrorCode2["MissingAuthorization"] = "MISSING_AUTHORIZATION";
  APIErrorCode2["InvalidAuthToken"] = "INVALID_AUTH_TOKEN";
  APIErrorCode2["InvalidToken"] = "INVALID_TOKEN";
  APIErrorCode2["TwoFactorRequired"] = "TWO_FACTOR_REQUIRED";
  APIErrorCode2["SudoModeRequired"] = "SUDO_MODE_REQUIRED";
  APIErrorCode2["NotFound"] = "NOT_FOUND";
  APIErrorCode2["UnknownUser"] = "UNKNOWN_USER";
  APIErrorCode2["UnknownGuild"] = "UNKNOWN_GUILD";
  APIErrorCode2["UnknownChannel"] = "UNKNOWN_CHANNEL";
  APIErrorCode2["UnknownMessage"] = "UNKNOWN_MESSAGE";
  APIErrorCode2["UnknownRole"] = "UNKNOWN_ROLE";
  APIErrorCode2["UnknownEmoji"] = "UNKNOWN_EMOJI";
  APIErrorCode2["UnknownSticker"] = "UNKNOWN_STICKER";
  APIErrorCode2["UnknownWebhook"] = "UNKNOWN_WEBHOOK";
  APIErrorCode2["UnknownInvite"] = "UNKNOWN_INVITE";
  APIErrorCode2["BadRequest"] = "BAD_REQUEST";
  APIErrorCode2["ValidationError"] = "VALIDATION_ERROR";
  APIErrorCode2["InvalidRequest"] = "INVALID_REQUEST";
  APIErrorCode2["InvalidFormBody"] = "INVALID_FORM_BODY";
  APIErrorCode2["RateLimited"] = "RATE_LIMITED";
  APIErrorCode2["SlowmodeRateLimited"] = "SLOWMODE_RATE_LIMITED";
  APIErrorCode2["InternalServerError"] = "INTERNAL_SERVER_ERROR";
  APIErrorCode2["BadGateway"] = "BAD_GATEWAY";
  APIErrorCode2["GatewayTimeout"] = "GATEWAY_TIMEOUT";
  APIErrorCode2["ServiceUnavailable"] = "SERVICE_UNAVAILABLE";
  APIErrorCode2["BotsCannotSendFriendRequests"] = "BOTS_CANNOT_SEND_FRIEND_REQUESTS";
  APIErrorCode2["BotAlreadyInGuild"] = "BOT_ALREADY_IN_GUILD";
  APIErrorCode2["BotApplicationNotFound"] = "BOT_APPLICATION_NOT_FOUND";
  APIErrorCode2["BotIsPrivate"] = "BOT_IS_PRIVATE";
  APIErrorCode2["NotABotApplication"] = "NOT_A_BOT_APPLICATION";
  APIErrorCode2["CannotSendEmptyMessage"] = "CANNOT_SEND_EMPTY_MESSAGE";
  APIErrorCode2["FileSizeTooLarge"] = "FILE_SIZE_TOO_LARGE";
  APIErrorCode2["MaxEmojis"] = "MAX_EMOJIS";
  APIErrorCode2["MaxStickers"] = "MAX_STICKERS";
  APIErrorCode2["MaxWebhooks"] = "MAX_WEBHOOKS";
  return APIErrorCode2;
})(APIErrorCode || {});

// src/gateway/opcodes.ts
var GatewayOpcodes = /* @__PURE__ */ ((GatewayOpcodes2) => {
  GatewayOpcodes2[GatewayOpcodes2["Dispatch"] = 0] = "Dispatch";
  GatewayOpcodes2[GatewayOpcodes2["Heartbeat"] = 1] = "Heartbeat";
  GatewayOpcodes2[GatewayOpcodes2["Identify"] = 2] = "Identify";
  GatewayOpcodes2[GatewayOpcodes2["PresenceUpdate"] = 3] = "PresenceUpdate";
  GatewayOpcodes2[GatewayOpcodes2["VoiceStateUpdate"] = 4] = "VoiceStateUpdate";
  GatewayOpcodes2[GatewayOpcodes2["Resume"] = 6] = "Resume";
  GatewayOpcodes2[GatewayOpcodes2["Reconnect"] = 7] = "Reconnect";
  GatewayOpcodes2[GatewayOpcodes2["RequestGuildMembers"] = 8] = "RequestGuildMembers";
  GatewayOpcodes2[GatewayOpcodes2["InvalidSession"] = 9] = "InvalidSession";
  GatewayOpcodes2[GatewayOpcodes2["Hello"] = 10] = "Hello";
  GatewayOpcodes2[GatewayOpcodes2["HeartbeatAck"] = 11] = "HeartbeatAck";
  return GatewayOpcodes2;
})(GatewayOpcodes || {});

// src/gateway/events.ts
var GatewayDispatchEvents = {
  // ─── Connection & Session ─────────────────────────────────────────────────
  /** Initial connection established; contains user, guilds, and session state. */
  Ready: "READY",
  /** Connection resumed after disconnect; sequence restored. */
  Resumed: "RESUMED",
  /** User's active sessions list replaced (e.g. new login). */
  SessionsReplace: "SESSIONS_REPLACE",
  // ─── User ──────────────────────────────────────────────────────────────────
  /** Current user's profile was updated. */
  UserUpdate: "USER_UPDATE",
  /** User settings changed (theme, locale, etc.). */
  UserSettingsUpdate: "USER_SETTINGS_UPDATE",
  /** Per-guild user settings changed (notifications, etc.). */
  UserGuildSettingsUpdate: "USER_GUILD_SETTINGS_UPDATE",
  /** Pinned DM order changed. */
  UserPinnedDmsUpdate: "USER_PINNED_DMS_UPDATE",
  /** Note on another user changed. */
  UserNoteUpdate: "USER_NOTE_UPDATE",
  /** Recent mention was cleared. */
  RecentMentionDelete: "RECENT_MENTION_DELETE",
  // ─── Saved Messages & Auth ─────────────────────────────────────────────────
  /** Message was saved (bookmarked). */
  SavedMessageCreate: "SAVED_MESSAGE_CREATE",
  /** Saved message was unsaved. */
  SavedMessageDelete: "SAVED_MESSAGE_DELETE",
  /** Auth session changed (login/logout on another client). */
  AuthSessionChange: "AUTH_SESSION_CHANGE",
  // ─── Presence ───────────────────────────────────────────────────────────────
  /** User's presence (status, activity) updated. */
  PresenceUpdate: "PRESENCE_UPDATE",
  // ─── Guild ─────────────────────────────────────────────────────────────────
  /** Bot joined a guild or guild data became available. */
  GuildCreate: "GUILD_CREATE",
  /** Guild was updated. */
  GuildUpdate: "GUILD_UPDATE",
  /** Bot left or was removed from a guild. */
  GuildDelete: "GUILD_DELETE",
  /** Member joined a guild. */
  GuildMemberAdd: "GUILD_MEMBER_ADD",
  /** Member was updated (roles, nickname, etc.). */
  GuildMemberUpdate: "GUILD_MEMBER_UPDATE",
  /** Member left or was removed from a guild. */
  GuildMemberRemove: "GUILD_MEMBER_REMOVE",
  /** Chunk of guild members (from request_guild_members). */
  GuildMembersChunk: "GUILD_MEMBERS_CHUNK",
  /** Member list view updated (lazy loading). */
  GuildMemberListUpdate: "GUILD_MEMBER_LIST_UPDATE",
  /** Guild sync state (passive/lazy loading). */
  GuildSync: "GUILD_SYNC",
  // ─── Roles ──────────────────────────────────────────────────────────────────
  /** Role was created. */
  GuildRoleCreate: "GUILD_ROLE_CREATE",
  /** Role was updated. */
  GuildRoleUpdate: "GUILD_ROLE_UPDATE",
  /** Multiple roles updated at once. */
  GuildRoleUpdateBulk: "GUILD_ROLE_UPDATE_BULK",
  /** Role was deleted. */
  GuildRoleDelete: "GUILD_ROLE_DELETE",
  // ─── Guild Assets ───────────────────────────────────────────────────────────
  /** Guild emojis changed. */
  GuildEmojisUpdate: "GUILD_EMOJIS_UPDATE",
  /** Guild stickers changed. */
  GuildStickersUpdate: "GUILD_STICKERS_UPDATE",
  // ─── Moderation ─────────────────────────────────────────────────────────────
  /** User was banned from a guild. */
  GuildBanAdd: "GUILD_BAN_ADD",
  /** User was unbanned from a guild. */
  GuildBanRemove: "GUILD_BAN_REMOVE",
  // ─── Channels ──────────────────────────────────────────────────────────────
  /** Channel was created. */
  ChannelCreate: "CHANNEL_CREATE",
  /** Channel was updated. */
  ChannelUpdate: "CHANNEL_UPDATE",
  /** Multiple channels updated at once. */
  ChannelUpdateBulk: "CHANNEL_UPDATE_BULK",
  /** Channel was deleted. */
  ChannelDelete: "CHANNEL_DELETE",
  /** Recipient added to group DM. */
  ChannelRecipientAdd: "CHANNEL_RECIPIENT_ADD",
  /** Recipient removed from group DM. */
  ChannelRecipientRemove: "CHANNEL_RECIPIENT_REMOVE",
  /** Pinned messages in a channel changed. */
  ChannelPinsUpdate: "CHANNEL_PINS_UPDATE",
  /** User acknowledged viewing pinned messages. */
  ChannelPinsAck: "CHANNEL_PINS_ACK",
  // ─── Internal / Passive ────────────────────────────────────────────────────
  /** Passive/lazy-loaded entity updates. */
  PassiveUpdates: "PASSIVE_UPDATES",
  // ─── Invites ───────────────────────────────────────────────────────────────
  /** Invite was created. */
  InviteCreate: "INVITE_CREATE",
  /** Invite was deleted or expired. */
  InviteDelete: "INVITE_DELETE",
  // ─── Messages ──────────────────────────────────────────────────────────────
  /** New message was sent. */
  MessageCreate: "MESSAGE_CREATE",
  /** Message was edited. */
  MessageUpdate: "MESSAGE_UPDATE",
  /** Message was deleted. */
  MessageDelete: "MESSAGE_DELETE",
  /** Multiple messages were deleted. */
  MessageDeleteBulk: "MESSAGE_DELETE_BULK",
  /** Reaction was added to a message. */
  MessageReactionAdd: "MESSAGE_REACTION_ADD",
  /** Reaction was removed from a message. */
  MessageReactionRemove: "MESSAGE_REACTION_REMOVE",
  /** All reactions were removed from a message. */
  MessageReactionRemoveAll: "MESSAGE_REACTION_REMOVE_ALL",
  /** All reactions of an emoji were removed from a message. */
  MessageReactionRemoveEmoji: "MESSAGE_REACTION_REMOVE_EMOJI",
  /** Message was acknowledged (read receipt). */
  MessageAck: "MESSAGE_ACK",
  // ─── Typing ─────────────────────────────────────────────────────────────────
  /** User started typing in a channel. */
  TypingStart: "TYPING_START",
  // ─── Webhooks ───────────────────────────────────────────────────────────────
  /** Webhooks for a channel were updated. */
  WebhooksUpdate: "WEBHOOKS_UPDATE",
  // ─── Relationships ──────────────────────────────────────────────────────────
  /** Relationship (friend, block) was added. */
  RelationshipAdd: "RELATIONSHIP_ADD",
  /** Relationship was updated. */
  RelationshipUpdate: "RELATIONSHIP_UPDATE",
  /** Relationship was removed. */
  RelationshipRemove: "RELATIONSHIP_REMOVE",
  // ─── Voice ───────────────────────────────────────────────────────────────────
  /** Voice state changed (join, leave, mute, etc.). */
  VoiceStateUpdate: "VOICE_STATE_UPDATE",
  /** Voice server allocation info (for connecting to voice). */
  VoiceServerUpdate: "VOICE_SERVER_UPDATE",
  // ─── Calls ──────────────────────────────────────────────────────────────────
  /** Call was created. */
  CallCreate: "CALL_CREATE",
  /** Call was updated. */
  CallUpdate: "CALL_UPDATE",
  /** Call was ended. */
  CallDelete: "CALL_DELETE",
  // ─── Favorites ──────────────────────────────────────────────────────────────
  /** Favorite meme/media was added. */
  FavoriteMemeCreate: "FAVORITE_MEME_CREATE",
  /** Favorite meme/media was updated. */
  FavoriteMemeUpdate: "FAVORITE_MEME_UPDATE",
  /** Favorite meme/media was removed. */
  FavoriteMemeDelete: "FAVORITE_MEME_DELETE",
  // ─── SDK / Compatibility (may come from API layer) ──────────────────────────
  /** Slash command or component interaction. */
  InteractionCreate: "INTERACTION_CREATE",
  /** Guild integrations changed. */
  GuildIntegrationsUpdate: "GUILD_INTEGRATIONS_UPDATE",
  /** Guild scheduled event was created. */
  GuildScheduledEventCreate: "GUILD_SCHEDULED_EVENT_CREATE",
  /** Guild scheduled event was updated. */
  GuildScheduledEventUpdate: "GUILD_SCHEDULED_EVENT_UPDATE",
  /** Guild scheduled event was cancelled. */
  GuildScheduledEventDelete: "GUILD_SCHEDULED_EVENT_DELETE"
};

// src/rest/routes.ts
var Routes = {
  // Channels
  channel: (id) => `/channels/${id}`,
  channelMessages: (id) => `/channels/${id}/messages`,
  channelMessage: (channelId, messageId) => `/channels/${channelId}/messages/${messageId}`,
  channelMessageReactions: (channelId, messageId) => `/channels/${channelId}/messages/${messageId}/reactions`,
  channelMessageReaction: (channelId, messageId, emoji) => `/channels/${channelId}/messages/${messageId}/reactions/${encodeURIComponent(emoji)}`,
  channelPins: (id) => `/channels/${id}/messages/pins`,
  /** Use channelPinMessage for PUT/DELETE pin operations. */
  channelPin: (channelId, messageId) => `/channels/${channelId}/messages/pins/${messageId}`,
  /** Pin/unpin: PUT or DELETE /channels/{id}/pins/{messageId}. Use for pin and unpin operations. */
  channelPinMessage: (channelId, messageId) => `/channels/${channelId}/pins/${messageId}`,
  channelBulkDelete: (id) => `/channels/${id}/messages/bulk-delete`,
  channelWebhooks: (id) => `/channels/${id}/webhooks`,
  channelTyping: (id) => `/channels/${id}/typing`,
  channelInvites: (id) => `/channels/${id}/invites`,
  channelPermission: (channelId, overwriteId) => `/channels/${channelId}/permissions/${overwriteId}`,
  channelRecipient: (channelId, userId) => `/channels/${channelId}/recipients/${userId}`,
  channelMessageAttachment: (channelId, messageId, attachmentId) => `/channels/${channelId}/messages/${messageId}/attachments/${attachmentId}`,
  // Guilds
  guilds: () => "/guilds",
  guild: (id) => `/guilds/${id}`,
  guildDelete: (guildId) => `/guilds/${guildId}/delete`,
  guildVanityUrl: (guildId) => `/guilds/${guildId}/vanity-url`,
  guildTextChannelFlexibleNames: (guildId) => `/guilds/${guildId}/text-channel-flexible-names`,
  guildDetachedBanner: (guildId) => `/guilds/${guildId}/detached-banner`,
  guildDisallowUnclaimedAccounts: (guildId) => `/guilds/${guildId}/disallow-unclaimed-accounts`,
  guildTransferOwnership: (guildId) => `/guilds/${guildId}/transfer-ownership`,
  guildRolesHoistPositions: (guildId) => `/guilds/${guildId}/roles/hoist-positions`,
  guildEmojisBulk: (guildId) => `/guilds/${guildId}/emojis/bulk`,
  guildStickersBulk: (guildId) => `/guilds/${guildId}/stickers/bulk`,
  guildChannels: (id) => `/guilds/${id}/channels`,
  guildMembers: (id) => `/guilds/${id}/members`,
  guildMember: (guildId, userId) => `/guilds/${guildId}/members/${userId}`,
  guildMemberRole: (guildId, userId, roleId) => `/guilds/${guildId}/members/${userId}/roles/${roleId}`,
  guildRoles: (id) => `/guilds/${id}/roles`,
  guildRole: (guildId, roleId) => `/guilds/${guildId}/roles/${roleId}`,
  guildBans: (id) => `/guilds/${id}/bans`,
  guildBan: (guildId, userId) => `/guilds/${guildId}/bans/${userId}`,
  guildInvites: (id) => `/guilds/${id}/invites`,
  invite: (code) => `/invites/${encodeURIComponent(code)}`,
  guildAuditLogs: (id) => `/guilds/${id}/audit-logs`,
  guildEmojis: (id) => `/guilds/${id}/emojis`,
  guildEmoji: (guildId, emojiId) => `/guilds/${guildId}/emojis/${emojiId}`,
  guildStickers: (id) => `/guilds/${id}/stickers`,
  guildSticker: (guildId, stickerId) => `/guilds/${guildId}/stickers/${stickerId}`,
  guildWebhooks: (id) => `/guilds/${id}/webhooks`,
  webhook: (id) => `/webhooks/${id}`,
  webhookExecute: (id, token) => `/webhooks/${id}/${token}`,
  // Users
  user: (id) => `/users/${id}`,
  currentUser: () => `/users/@me`,
  currentUserGuilds: () => `/users/@me/guilds`,
  leaveGuild: (guildId) => `/users/@me/guilds/${guildId}`,
  userMeChannels: () => `/users/@me/channels`,
  /** GET /users/{id}/profile. Pass guildId for server-specific profile. */
  userProfile: (id, guildId) => guildId ? `/users/${id}/profile?guild_id=${guildId}` : `/users/${id}/profile`,
  // Instance (unauthenticated)
  instance: () => "/instance",
  // Gateway
  gatewayBot: () => `/gateway/bot`,
  // Streams (voice channel screen share preview)
  streamPreview: (streamKey) => `/streams/${encodeURIComponent(streamKey)}/preview`,
  // Application commands (slash commands)
  applicationCommands: (applicationId) => `/applications/${applicationId}/commands`,
  applicationCommand: (applicationId, commandId) => `/applications/${applicationId}/commands/${commandId}`,
  interactionCallback: (interactionId, interactionToken) => `/interactions/${interactionId}/${interactionToken}/callback`,
  // OAuth2 / Bot
  oauth2ApplicationBot: (id) => `/oauth2/applications/${id}/bot`,
  oauth2ApplicationBotResetToken: (id) => `/oauth2/applications/${id}/bot/reset-token`
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  APIErrorCode,
  ChannelType,
  DefaultMessageNotifications,
  EmbedMediaFlags,
  FLUXER_EPOCH,
  GatewayDispatchEvents,
  GatewayOpcodes,
  GuildExplicitContentFilter,
  GuildMFALevel,
  GuildVerificationLevel,
  MessageAttachmentFlags,
  MessageFlags,
  MessageType,
  OverwriteType,
  Routes,
  WebhookType
});
