// @fluxerjs/builders/dist/index.mjs
// src/messages/EmbedBuilder.ts
import { resolveColor } from "@fluxerjs/util";
var EMBED_MAX = {
  title: 256,
  description: 4096,
  fields: 25,
  fieldName: 256,
  fieldValue: 1024,
  footerText: 2048,
  authorName: 256,
  total: 6e3
};
function toEmbedMedia(input) {
  if (typeof input === "string") {
    return { url: input };
  }
  if (!URL.canParse(input.url)) {
    throw new Error("Invalid embed media URL");
  }
  const media = { url: input.url };
  if (input.content_type != null) media.content_type = input.content_type;
  if (input.width != null) media.width = input.width;
  if (input.height != null) media.height = input.height;
  if (input.description != null) media.description = input.description;
  if (input.placeholder != null) media.placeholder = input.placeholder;
  if (input.duration != null) media.duration = input.duration;
  if (input.flags != null) media.flags = input.flags;
  return media;
}
var EmbedBuilder = class _EmbedBuilder {
  data = {};
  /** Set the embed title. Max 256 characters. */
  setTitle(title) {
    if (title !== null && title.length > EMBED_MAX.title)
      throw new RangeError(`Title must be \u2264${EMBED_MAX.title} characters`);
    this.data.title = title ?? void 0;
    return this;
  }
  /** Set the embed description. Max 4096 characters. */
  setDescription(description) {
    if (description !== null && description.length > EMBED_MAX.description)
      throw new RangeError(`Description must be \u2264${EMBED_MAX.description} characters`);
    this.data.description = description ?? void 0;
    return this;
  }
  /** Set the embed URL (title becomes a link). */
  setURL(url) {
    if (url != null && url !== "" && !URL.canParse(url)) {
      throw new Error("Invalid embed URL");
    }
    this.data.url = url ?? void 0;
    return this;
  }
  /** Set the embed color. Number (hex), hex string, or `[r,g,b]` array. */
  setColor(color) {
    if (color === null) {
      this.data.color = void 0;
      return this;
    }
    this.data.color = typeof color === "number" ? color : resolveColor(color);
    return this;
  }
  /** Set the embed timestamp. Omit for current time. */
  setTimestamp(timestamp) {
    if (timestamp === void 0 || timestamp === null) {
      this.data.timestamp = void 0;
      return this;
    }
    const date = timestamp instanceof Date ? timestamp : new Date(timestamp);
    this.data.timestamp = date.toISOString();
    return this;
  }
  /** Set the embed author (name, optional icon URL and link). */
  setAuthor(options) {
    if (!options) {
      this.data.author = void 0;
      return this;
    }
    const name = options.name.slice(0, EMBED_MAX.authorName);
    const author = { name };
    if (options.url) author.url = options.url;
    if (options.iconURL) author.icon_url = options.iconURL;
    this.data.author = author;
    return this;
  }
  /** Set the embed footer (text, optional icon URL). */
  setFooter(options) {
    if (!options) {
      this.data.footer = void 0;
      return this;
    }
    const text = options.text.slice(0, EMBED_MAX.footerText);
    const footer = { text };
    if (options.iconURL) footer.icon_url = options.iconURL;
    this.data.footer = footer;
    return this;
  }
  /** Set the embed image (URL string or full media options). */
  setImage(input) {
    this.data.image = input ? toEmbedMedia(input) : void 0;
    return this;
  }
  /** Set the embed thumbnail (URL string or full media options). */
  setThumbnail(input) {
    this.data.thumbnail = input ? toEmbedMedia(input) : void 0;
    return this;
  }
  /**
   * Set the embed video. Supported by Fluxer.
   * Embed stays type 'rich'; this adds the .video field.
   * Include a title (e.g. setTitle) when using video.
   *
   * @param input - Video URL, full media options (e.g. duration for progress bars), or null to clear
   */
  setVideo(input) {
    this.data.video = input ? toEmbedMedia(input) : void 0;
    return this;
  }
  /**
   * Set the embed audio. Supported by Fluxer.
   *
   * @param input - Audio URL, full media options, or null to clear
   */
  setAudio(input) {
    this.data.audio = input ? toEmbedMedia(input) : void 0;
    return this;
  }
  /** Add one or more fields. Max 25 fields. */
  addFields(...fields) {
    const current = (this.data.fields ?? []).slice();
    for (const f of fields) {
      if (current.length >= EMBED_MAX.fields) break;
      current.push({
        name: f.name.slice(0, EMBED_MAX.fieldName),
        value: f.value.slice(0, EMBED_MAX.fieldValue),
        inline: f.inline
      });
    }
    this.data.fields = current.length ? current : void 0;
    return this;
  }
  spliceFields(index, deleteCount, ...fields) {
    const current = (this.data.fields ?? []).slice();
    const toAdd = fields.map((f) => ({
      name: f.name.slice(0, EMBED_MAX.fieldName),
      value: f.value.slice(0, EMBED_MAX.fieldValue),
      inline: f.inline
    }));
    current.splice(index, deleteCount, ...toAdd);
    this.data.fields = current.length ? current : void 0;
    return this;
  }
  /** Convert to API embed format for `reply`, `send`, or `edit`. */
  toJSON() {
    const totalLength = [
      this.data.title,
      this.data.description,
      ...(this.data.fields ?? []).flatMap((f) => [f.name, f.value]),
      this.data.footer?.text
    ].filter(Boolean).join("").length;
    if (totalLength > EMBED_MAX.total)
      throw new RangeError(`Embed total length must be \u2264${EMBED_MAX.total}`);
    return { ...this.data, type: "rich" };
  }
  /** Create an EmbedBuilder from an existing API embed. */
  static from(data) {
    const b = new _EmbedBuilder();
    b.data.title = data.title ?? void 0;
    b.data.description = data.description ?? void 0;
    b.data.url = data.url ?? void 0;
    b.data.color = data.color ?? void 0;
    b.data.timestamp = data.timestamp ?? void 0;
    b.data.author = data.author ?? void 0;
    b.data.footer = data.footer ?? void 0;
    b.data.image = data.image ?? void 0;
    b.data.thumbnail = data.thumbnail ?? void 0;
    b.data.video = data.video ?? void 0;
    b.data.audio = data.audio ?? void 0;
    b.data.fields = data.fields ?? void 0;
    return b;
  }
};

// src/messages/AttachmentBuilder.ts
var AttachmentBuilder = class {
  id;
  filename;
  description;
  spoiler;
  /** @param id - Index of the attachment (0-based). Must match the FormData part order. */
  constructor(id, filename, options) {
    if (!filename?.trim()) {
      throw new Error("Filename is required");
    }
    this.id = id;
    this.filename = options?.spoiler ? `SPOILER_${filename}` : filename;
    this.description = options?.description ?? void 0;
    this.spoiler = options?.spoiler ?? false;
  }
  /** Set the displayed filename. */
  setName(name) {
    if (!name?.trim()) {
      throw new Error("Filename is required");
    }
    this.filename = this.spoiler ? `SPOILER_${name}` : name;
    return this;
  }
  /** Set the attachment description (alt text). */
  setDescription(description) {
    this.description = description ?? void 0;
    return this;
  }
  /** Mark the attachment as a spoiler (blurred until clicked). */
  setSpoiler(spoiler = true) {
    this.spoiler = spoiler;
    if (spoiler && !this.filename.startsWith("SPOILER_"))
      this.filename = `SPOILER_${this.filename}`;
    else if (!spoiler && this.filename.startsWith("SPOILER_"))
      this.filename = this.filename.slice(8);
    return this;
  }
  /** Convert to API format for MessagePayload. */
  toJSON() {
    return {
      id: this.id,
      filename: this.filename,
      description: this.description ?? void 0
    };
  }
};

// src/messages/MessagePayload.ts
var CONTENT_MAX = 2e3;
var EMBEDS_MAX = 10;
var MessagePayload = class _MessagePayload {
  static ContentMaxLength = CONTENT_MAX;
  data = {};
  /** Set message text. Max 2000 characters. */
  setContent(content) {
    if (content !== null && content.length > CONTENT_MAX)
      throw new RangeError(`Content must be \u2264${CONTENT_MAX} characters`);
    this.data.content = content ?? void 0;
    return this;
  }
  /** Set embeds. Max 10. Replaces existing. */
  setEmbeds(embeds) {
    if (!embeds?.length) {
      this.data.embeds = void 0;
      return this;
    }
    if (embeds.length > EMBEDS_MAX) throw new RangeError(`Embeds must be \u2264${EMBEDS_MAX}`);
    this.data.embeds = embeds.map((e) => e instanceof EmbedBuilder ? e.toJSON() : e);
    return this;
  }
  /** Add one embed. Max 10 total. */
  addEmbed(embed) {
    const list = (this.data.embeds ?? []).slice();
    if (list.length >= EMBEDS_MAX) throw new RangeError(`Embeds must be \u2264${EMBEDS_MAX}`);
    list.push(embed instanceof EmbedBuilder ? embed.toJSON() : embed);
    this.data.embeds = list;
    return this;
  }
  /** Set attachment metadata (for files sent with the request). */
  setAttachments(attachments) {
    if (!attachments?.length) {
      this.data.attachments = void 0;
      return this;
    }
    this.data.attachments = attachments.map(
      (a) => a instanceof AttachmentBuilder ? a.toJSON() : a
    );
    return this;
  }
  /** Set reply reference (creates a reply to another message). */
  setReply(reference) {
    if (!reference) {
      this.data.message_reference = void 0;
      return this;
    }
    this.data.message_reference = {
      channel_id: reference.channel_id,
      message_id: reference.message_id,
      guild_id: reference.guild_id ?? void 0
    };
    return this;
  }
  /** Enable text-to-speech. */
  setTTS(tts) {
    this.data.tts = tts;
    return this;
  }
  /** Set message flags (e.g. ephemeral, suppress embeds). */
  setFlags(flags) {
    this.data.flags = flags;
    return this;
  }
  /** Get the payload as a plain object. */
  toJSON() {
    return { ...this.data };
  }
  /** Create a MessagePayload from a string or options object. */
  static create(contentOrOptions) {
    const payload = new _MessagePayload();
    if (typeof contentOrOptions === "string") {
      payload.setContent(contentOrOptions);
    } else if (contentOrOptions && typeof contentOrOptions === "object") {
      if (contentOrOptions.content !== void 0)
        payload.setContent(contentOrOptions.content ?? null);
      if (contentOrOptions.embeds?.length) payload.setEmbeds(contentOrOptions.embeds);
      if (contentOrOptions.attachments?.length)
        payload.setAttachments(contentOrOptions.attachments);
      if (contentOrOptions.message_reference)
        payload.setReply(contentOrOptions.message_reference);
      if (contentOrOptions.tts !== void 0) payload.setTTS(contentOrOptions.tts);
      if (contentOrOptions.flags !== void 0) payload.setFlags(contentOrOptions.flags);
    }
    return payload;
  }
};
export {
  AttachmentBuilder,
  EmbedBuilder,
  MessagePayload
};
